src/hotspot/share/opto/castnode.cpp

```diff
@@ -239,11 +233,11 @@ Node *CastIINode::Ideal(PhaseGVN *phase, bool can_reshape) {
   if (progress != nullptr) {
     return progress;
   }
-  if (can_reshape && !_range_check_dependency && !phase->C->post_loop_opts_phase()) {
+  if (can_reshape && !phase->C->post_loop_opts_phase()) {
     // makes sure we run ::Value to potentially remove type assertion after loop opts
     phase->C->record_for_post_loop_opts_igvn(this);
   }
-  if (!_range_check_dependency) {
+  if (!_type->is_int()->empty()) {
```

Comment:
Can you also explain this change, please?

Reply:
It's similar to the case above. `ConvI2L` has a similar transformation (`Add` node pushed through `ConvI2L`). For range checks, the `CastII` gets in the way but it wasn't an issue without the patch I propose (the `CastII` ends up going away). Leaving the `CastII` requires that the transformation happens on the `CastII` as well. The:

```
!_type->is_int()->empty()
```
test doesn't replace the
```
!_range_check_dependency
```
 I removed the second one, ran some testing and then had a failure where the `ConstraintCastNode::optimize_integer_cast()` crashes because it gets called with an empty type so I added the test for an empty type. I don't remember the details of the failure other than it seemed like a corner case.